### PR TITLE
planner: update the doc of SPM

### PR DESCRIPTION
**Pull request by qw4990**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.2 (TiDB 7.2 versions)
- [x] v7.1 (TiDB 7.1 versions)
- [ ] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [x] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by elsa0520 — APPROVED** [review]

File: sql-plan-management.md [hunk 1]
```diff
@@ -177,10 +177,11 @@ CREATE BINDING FOR SELECT * FROM t WHERE a > 1 USING SELECT * FROM t use index(i
 
 如需将 SQL 语句的执行计划固定为之前使用过的执行计划，可以使用 `plan_digest` 为该 SQL 语句绑定一个历史的执行计划。相比于使用 SQL 创建绑定的方式，此方式更加简便。
 
-目前，根据历史执行计划创建绑定有一些限制：
+目前，根据历史执行计划创建绑定有一些说明：
```
Comment (Oreoxmt) [inline sql-plan-management.md:180]:
```suggestion
以下为根据历史执行计划创建绑定的说明：
```

File: sql-plan-management.md [hunk 2]
```diff
@@ -177,10 +177,11 @@ CREATE BINDING FOR SELECT * FROM t WHERE a > 1 USING SELECT * FROM t use index(i
...
 
 - 该功能是根据历史的执行计划生成 hint 而实现的绑定，历史的执行计划来源是 [Statement Summary Tables](/statement-summary-tables.md)，因此在使用此功能之前需开启系统变量 [`tidb_enable_stmt_summary`](/system-variables.md#tidb_enable_stmt_summary-从-v304-版本开始引入)。
-- 对于带有子查询的查询、访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 对于访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 原计划对应 SQL 中的 hint 也会被一并用在创建的绑定中，如执行 `select /*+ max_execution_time(1000) */ * from t` 后，使用其 `plan_digest` 创建的绑定中会带上 `max_execution_time(1000)`。
```
Comment (Oreoxmt) [inline sql-plan-management.md:184]:
```suggestion
- 原计划对应 SQL 中的 hint 也会被一并用在创建的绑定中，如执行 `SELECT /*+ max_execution_time(1000) */ * FROM t` 后，使用其 `plan_digest` 创建的绑定中会带上 `max_execution_time(1000)`。
```

File: optimizer-hints.md [hunk 3]
```diff
@@ -814,3 +786,78 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+
+## 常见 Hint 不生效问题排查
+
+### MySQL 命令行客户端清除 Hint 导致不生效
+
+MySQL 命令行客户端在 5.7.7 版本之前默认清除了 Optimizer Hints。如果需要在这些早期版本的客户端中使用 `Hint` 语法，需要在启动客户端时加上 `--comments` 选项，例如 `mysql -h 127.0.0.1 -P 4000 -uroot --comments`。
+
+### 不指定库名导致 Hint 不生效
+
+如果创建链接时未指定数据库名，则可能出现 Hint 失效的情况。例如：
```
Comment (Oreoxmt) [inline optimizer-hints.md:798]:
```suggestion
如果创建连接时未指定数据库名，则可能出现 Hint 失效的情况。例如：
```

File: optimizer-hints.md [hunk 4]
```diff
@@ -814,3 +786,78 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+mysql> show warnings;
++---------+------+----------------------------------------------------------------------+
+| Level   | Code | Message                                                              |
++---------+------+----------------------------------------------------------------------+
+| Warning | 1815 | use_index(.t, a) is inapplicable, check whether the table(.t) exists |
++---------+------+----------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+由于无法识别表 `t` 对应的数据库名，因此 `use_index(t, a)` 无法生效。
```
Comment (Oreoxmt) [inline optimizer-hints.md:815]:
````suggestion
使用 `mysql -h127.0.0.1 -P4000 -uroot` 命令连接数据库时，未使用 `-D` 参数指定数据库名。然后执行下面的 SQL：

```sql
SELECT /*+ use_index(t, a) */ a FROM test.t;
SHOW WARNINGS;
```

由于无法识别表 `t` 对应的数据库名，因此 `use_index(t, a)` Hint 无法生效。

```sql
+---------+------+----------------------------------------------------------------------+
| Level   | Code | Message                                                              |
+---------+------+----------------------------------------------------------------------+
| Warning | 1815 | use_index(.t, a) is inapplicable, check whether the table(.t) exists |
+---------+------+----------------------------------------------------------------------+
1 row in set (0.00 sec)
```
````

File: optimizer-hints.md [hunk 5]
```diff
@@ -814,3 +786,78 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+mysql> show warnings;
++---------+------+----------------------------------------------------------------------------------+
+| Level   | Code | Message                                                                          |
++---------+------+----------------------------------------------------------------------------------+
+| Warning | 1815 | use_index(test2.t1, a) is inapplicable, check whether the table(test2.t1) exists |
++---------+------+----------------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+由于 `t1` 不在当前数据库 `test2` 下，因此 Hint `use_index(t1, a)` 无法正确的识别，需要显示加上库名 `use_index(test1.t1, a)`。
```
Comment (Oreoxmt) [inline optimizer-hints.md:844]:
````suggestion
对于跨库查询中需要访问的表，需要显式地指定数据库名，否则可能出现 Hint 失效的情况。例如执行下面跨库查询的 SQL 语句：

```sql
USE test1;
CREATE TABLE t1(a INT, KEY(a));
USE test2;
CREATE TABLE t2(a INT, KEY(a));
SELECT /*+ use_index(t1, a) */ * FROM test1.t1, t2;
SHOW WARNINGS;
```

由于 `t1` 不在当前数据库 `test2` 下，因此 `use_index(t1, a)` Hint 无法被正确地识别。此时，需要显式地指定库名，即修改为 `use_index(test1.t1, a)`。

```sql
+---------+------+----------------------------------------------------------------------------------+
| Level   | Code | Message                                                                          |
+---------+------+----------------------------------------------------------------------------------+
| Warning | 1815 | use_index(test2.t1, a) is inapplicable, check whether the table(test2.t1) exists |
+---------+------+----------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
````

File: optimizer-hints.md [hunk 6]
```diff
@@ -814,3 +786,78 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Level   | Code | Message                                                                                                                                                                                                                 |
++---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| Warning | 1064 | You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:8066]Optimizer hint can only be followed by certain keywords like SELECT, INSERT, etc. |
++---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+1 row in set (0.01 sec)
+```
```
Comment (Oreoxmt) [inline optimizer-hints.md:861]:
````suggestion
如果没有将 Hint 正确地放在指定关键字的后面，它将无法生效。例如：

```sql
SELECT * /*+ use_index(t, a) */ FROM t;
SHOW WARNINGS;
```

Warning 信息如下：

```sql
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                                                                                 |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1064 | You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:8066]Optimizer hint can only be followed by certain keywords like SELECT, INSERT, etc. |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```
````

File: optimizer-hints.md [hunk 7]
```diff
@@ -814,3 +786,78 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+
+具体见 [Hint 语法](#语法)部分。
```
Comment (Oreoxmt) [inline optimizer-hints.md:863]:
```suggestion
具体的语法规则参见 [Hint 语法](#语法)部分。
```

File: optimizer-hints.md [hunk 8]
```diff
@@ -814,3 +786,77 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+
+### MySQL 命令行客户端清除 Hint 导致不生效
+
+MySQL 命令行客户端在 5.7.7 版本之前默认清除了 Optimizer Hints。如果需要在这些早期版本的客户端中使用 Hint 语法，需要在启动客户端时加上 `--comments` 选项。例如 `mysql -h 127.0.0.1 -P 4000 -uroot --comments`。
+
+### 不指定库名导致 Hint 不生效
+
+如果创建连接时未指定数据库名，则可能出现 Hint 失效的情况。例如：
+
+使用 `mysql -h127.0.0.1 -P4000 -uroot` 命令连接数据库时，未使用 `-D` 参数指定数据库名。然后执行下面的 SQL：
```
Comment (qiancai) [inline optimizer-hints.md:800]:
```suggestion
使用 `mysql -h127.0.0.1 -P4000 -uroot` 命令连接数据库时，未使用 `-D` 参数指定数据库名。然后执行下面的 SQL 语句：
```

File: optimizer-hints.md [hunk 9]
```diff
@@ -814,3 +786,77 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Level   | Code | Message                                                                          |
++---------+------+----------------------------------------------------------------------------------+
+| Warning | 1815 | use_index(test2.t1, a) is inapplicable, check whether the table(test2.t1) exists |
++---------+------+----------------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+### Hint 位置不正确导致不生效
+
+如果没有将 Hint 正确地放在指定关键字的后面，它将无法生效。例如：
```
Comment (qiancai) [inline optimizer-hints.md:844]:
```suggestion
如果没有按照 Optimizer Hints  语法将 Hint 正确地放在指定关键字的后面，它将无法生效。例如：
```

File: sql-plan-management.md [hunk 10]
```diff
@@ -177,10 +177,11 @@ CREATE BINDING FOR SELECT * FROM t WHERE a > 1 USING SELECT * FROM t use index(i
 
 如需将 SQL 语句的执行计划固定为之前使用过的执行计划，可以使用 `plan_digest` 为该 SQL 语句绑定一个历史的执行计划。相比于使用 SQL 创建绑定的方式，此方式更加简便。
 
-目前，根据历史执行计划创建绑定有一些限制：
+以下为根据历史执行计划创建绑定的说明：
 
 - 该功能是根据历史的执行计划生成 hint 而实现的绑定，历史的执行计划来源是 [Statement Summary Tables](/statement-summary-tables.md)，因此在使用此功能之前需开启系统变量 [`tidb_enable_stmt_summary`](/system-variables.md#tidb_enable_stmt_summary-从-v304-版本开始引入)。
-- 对于带有子查询的查询、访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 对于访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 原计划对应 SQL 中的 hint 也会被一并用在创建的绑定中，如执行 `SELECT /*+ max_execution_time(1000) */ * FROM t` 后，使用其 `plan_digest` 创建的绑定中会带上 `max_execution_time(1000)`。
```
Comment (qiancai) [inline sql-plan-management.md:184]:
```suggestion
- 原执行计划对应 SQL 语句中的 hint 也会被应用在创建的绑定中，如执行 `SELECT /*+ max_execution_time(1000) */ * FROM t` 后，使用其 `plan_digest` 创建的绑定中会带上 `max_execution_time(1000)`。
```

File: sql-plan-management.md [hunk 11]
```diff
@@ -177,10 +177,12 @@ CREATE BINDING FOR SELECT * FROM t WHERE a > 1 USING SELECT * FROM t use index(i
...
 如需将 SQL 语句的执行计划固定为之前使用过的执行计划，可以使用 `plan_digest` 为该 SQL 语句绑定一个历史的执行计划。相比于使用 SQL 创建绑定的方式，此方式更加简便。
 
-目前，根据历史执行计划创建绑定有一些限制：
+以下为根据历史执行计划创建绑定的注意事项：
 
 - 该功能是根据历史的执行计划生成 hint 而实现的绑定，历史的执行计划来源是 [Statement Summary Tables](/statement-summary-tables.md)，因此在使用此功能之前需开启系统变量 [`tidb_enable_stmt_summary`](/system-variables.md#tidb_enable_stmt_summary-从-v304-版本开始引入)。
-- 对于带有子查询的查询、访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 对于访问 TiFlash 的查询、3 张表或更多表进行 Join 的查询，目前还不支持通过历史执行计划进行绑定。
+- 原执行计划对应 SQL 语句中的 hint 也会被应用在创建的绑定中，如执行 `SELECT /*+ max_execution_time(1000) */ * FROM t` 后，使用其 `plan_digest` 创建的绑定中会带上 `max_execution_time(1000)`。
+- 目前，如果原本的查询包含非关联子查询，由于非关联子查询会被提前执行，其对应的 Hint 可能会有所缺失，需要注意。
```
Comment (Oreoxmt) [inline sql-plan-management.md:185]:
```suggestion
- 需要注意的是，如果原查询中包含非关联子查询，由于非关联子查询会被提前执行，其对应的 Hint 可能会缺失。
```

File: optimizer-hints.md [hunk 12]
```diff
@@ -814,3 +786,111 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+mysql> show warnings;
++---------+------+----------------------------------------------------------------------------+
+| Level   | Code | Message                                                                    |
++---------+------+----------------------------------------------------------------------------+
+| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(t1) */ or /*+ TIDB_INLJ(t1) */ is inapplicable |
++---------+------+----------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+上面的例子中 `t1.k` 和 `t2.k` 的 Collation 设置不同（分别为 `utf8mb4_general_ci` 和 `utf8mb4_bin`），导致 IndexJoin 无法适用，因此 `INL_JOIN` 或者 `TIDB_INLJ` Hint 也会无法生效。
```
Comment (Oreoxmt) [inline optimizer-hints.md:896]:
````suggestion
### 排序规则不同导致 `INL_JOIN` Hint 不生效

如果两个表的 Join key 的排序规则设置不同，将无法使用 IndexJoin 来执行查询。此时 `INL_JOIN` Hint 将无法生效。例如：

```sql
CREATE TABLE t1 (k varchar(8), key(k)) COLLATE=utf8mb4_general_ci;
CREATE TABLE t2 (k varchar(8), key(k)) COLLATE=utf8mb4_bin;
EXPLAIN SELECT /*+ tidb_inlj(t1) */ * FROM t1, t2 WHERE t1.k=t2.k;
```

查询计划输出结果如下：

```sql
+-----------------------------+----------+-----------+----------------------+----------------------------------------------+
| id                          | estRows  | task      | access object        | operator info                                |
+-----------------------------+----------+-----------+----------------------+----------------------------------------------+
| HashJoin_19                 | 12487.50 | root      |                      | inner join, equal:[eq(test.t1.k, test.t2.k)] |
| ├─IndexReader_24(Build)     | 9990.00  | root      |                      | index:IndexFullScan_23                       |
| │ └─IndexFullScan_23        | 9990.00  | cop[tikv] | table:t2, index:k(k) | keep order:false, stats:pseudo               |
| └─IndexReader_22(Probe)     | 9990.00  | root      |                      | index:IndexFullScan_21                       |
|   └─IndexFullScan_21        | 9990.00  | cop[tikv] | table:t1, index:k(k) | keep order:false, stats:pseudo               |
+-----------------------------+----------+-----------+----------------------+----------------------------------------------+
5 rows in set, 1 warning (0.00 sec)
```

上面的 SQL 语句中 `t1.k` 和 `t2.k` 的排序规则不同（分别为 `utf8mb4_general_ci` 和 `utf8mb4_bin`），导致 IndexJoin 无法适用。因此 `INL_JOIN` 或 `TIDB_INLJ` Hint 也无法生效。

```sql
SHOW WARNINGS;
+---------+------+----------------------------------------------------------------------------+
| Level   | Code | Message                                                                    |
+---------+------+----------------------------------------------------------------------------+
| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(t1) */ or /*+ TIDB_INLJ(t1) */ is inapplicable |
+---------+------+----------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
````

File: optimizer-hints.md [hunk 13]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Warning | 1064 | You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:8066]Optimizer hint can only be followed by certain keywords like SELECT, INSERT, etc. |
++---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+1 row in set (0.01 sec)
+```
+
+在上面的示例中，你需要将 Hint 直接放在 `SELECT` 关键字之后。具体的语法规则参见 [Hint 语法](#语法)部分。
+
+### 排序规则不同导致 `INL_JOIN` Hint 不生效
+
+如果两个表的 Join key 的排序规则设置不同，将无法使用 IndexJoin 来执行查询。此时 `INL_JOIN` Hint 将无法生效。例如：
```
Comment (Oreoxmt) [inline optimizer-hints.md:866]:
```suggestion
如果两个表的 Join key 的排序规则设置不同，将无法使用 IndexJoin 来执行查询。此时 [`INL_JOIN` Hint](#inl_joint1_name--tl_name-) 将无法生效。例如：
```

File: optimizer-hints.md [hunk 14]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Level   | Code | Message                                                                    |
++---------+------+----------------------------------------------------------------------------+
+| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(t1) */ or /*+ TIDB_INLJ(t1) */ is inapplicable |
++---------+------+----------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+### 连接顺序导致 `INL_JOIN` Hint 不生效
+
+`INL_JOIN(t1, t2)` 或 `TIDB_INLJ(t1, t2)` 的语义是让 `t1` 和 `t2` 在与其他表连接时使用 `IndexJoin`，而不是用 `IndexJoin` 直接连接 `t1` 和 `t2`。例如：
```
Comment (qw4990) [inline optimizer-hints.md:903]:
```suggestion
[`INL_JOIN(t1, t2)`](#inl_joint1_name--tl_name-) 或 `TIDB_INLJ(t1, t2)` 的语义是让 `t1` 和 `t2` 在与其他表连接时使用 `IndexJoin`，而不是直接将 `t1` 和 `t2` 进行 `IndexJoin` 连接。例如：
```

File: optimizer-hints.md [hunk 15]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| │   └─TableRangeScan_30         | 10000.00 | cop[tikv] | table:t1      | range: decided by [test.t2.id], keep order:false, stats:pseudo                                                                                                          |
+| └─TableReader_13(Probe)         | 12500.00 | root      |               | data:TableRangeScan_12                                                                                                                                                  |
+|   └─TableRangeScan_12           | 12500.00 | cop[tikv] | table:t3      | range: decided by [test.t2.id test.t1.id], keep order:false, stats:pseudo                                                                                               |
++---------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+在上面例子中，`t1` 和 `t3` 并没有直接被一个 `IndexJoin` 连接起来。
+
+如果想要直接使用 `IndexJoin` 来连接 `t1` 和 `t3`，需要先使用 `LEADING(t1, t3)` Hint 指定连接顺序，让 `t1` 和 `t3` 会被直接进行连接，再配合 `INL_JOIN`。例如：
```
Comment (qw4990) [inline optimizer-hints.md:923]:
```suggestion
如果想要直接使用 `IndexJoin` 来连接 `t1` 和 `t3`，需要先使用 [`LEADING` Hint](#leadingt1_name--tl_name-) 指定 `t1` 和 `t3` 的连接顺序，然后再配合使用 `INL_JION`。例如：
```

File: optimizer-hints.md [hunk 16]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Level   | Code | Message                                                                    |
++---------+------+----------------------------------------------------------------------------+
+| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(t1) */ or /*+ TIDB_INLJ(t1) */ is inapplicable |
++---------+------+----------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+### 连接顺序导致 `INL_JOIN` Hint 不生效
+
+[`INL_JOIN(t1, t2)`](#inl_joint1_name--tl_name-) 或 `TIDB_INLJ(t1, t2)` 的语义是让 `t1` 和 `t2` 在与其他表连接时使用 `IndexJoin`，而不是直接将 `t1` 和 `t2` 进行 `IndexJoin` 连接。例如：
```
Comment (time-and-fate) [inline optimizer-hints.md:903]: 这里感觉可以稍微严谨点吧，`inl_join` 是指定表作为 inner 表参与 index join，和“与其他表连接时使用 `IndexJoin`”略有不同。

File: optimizer-hints.md [hunk 17]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+
+```sql
+mysql> explain select /*+ tidb_inlj(t1, t3) */ * from t1, t2, t3 where t1.id=t2.id and t2.id=t3.id and t1.id=t3.id;
```
Comment (time-and-fate) [inline optimizer-hints.md:906]: 前面文档已经说推荐 `inl_join` 了，感觉统一用 `inl_join` 比较好吧。
```suggestion
EXPLAIN SELECT /*+ inl_join(t1, t3) */ * FROM t1, t2, t3 WHERE t1.id = t2.id AND t2.id = t3.id AND t1.id = t3.id;
```

File: optimizer-hints.md [hunk 18]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+|   └─TableRangeScan_12           | 12500.00 | cop[tikv] | table:t3      | range: decided by [test.t2.id test.t1.id], keep order:false, stats:pseudo                                                                                               |
++---------------------------------+----------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+在上面例子中，`t1` 和 `t3` 并没有直接被一个 `IndexJoin` 连接起来。
+
+如果想要直接使用 `IndexJoin` 来连接 `t1` 和 `t3`，需要先使用 [`LEADING` Hint](#leadingt1_name--tl_name-) 指定 `t1` 和 `t3` 的连接顺序，然后再配合使用 `INL_JION`。例如：
+
+```sql
+mysql> explain select /*+ leading(t1, t3), tidb_inlj(t3) */ * from t1, t2, t3 where t1.id=t2.id and t2.id=t3.id and t1.id=t3.id;
```
Comment (time-and-fate) [inline optimizer-hints.md:926]:
```suggestion
EXPLAIN SELECT /*+ leading(t1, t3), inl_join(t3) */ * FROM t1, t2, t3 WHERE t1.id = t2.id AND t2.id = t3.id AND t1.id = t3.id;
```

File: optimizer-hints.md [hunk 19]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+```sql
++---------+------+----------------------------------------------------------------------+
+| Level   | Code | Message                                                              |
++---------+------+----------------------------------------------------------------------+
+| Warning | 1815 | use_index(.t, a) is inapplicable, check whether the table(.t) exists |
++---------+------+----------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+### 跨库查询不指定库名导致 Hint 不生效
```
[the remark this answers is not in the document]
Comment (qw4990) [reply optimizer-hints.md:818]: 嗯，原理是一样，但是从用户角度看造成问题的方式不一样，分两个让用户自己对号入座吧

File: optimizer-hints.md (hunk 16 above)
Comment (qw4990) [reply optimizer-hints.md]: 已更新

File: optimizer-hints.md [hunk 20]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+| Warning | 1064 | You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use [parser:8066]Optimizer hint can only be followed by certain keywords like SELECT, INSERT, etc. |
++---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+1 row in set (0.01 sec)
+```
+
+在上面的示例中，你需要将 Hint 直接放在 `SELECT` 关键字之后。具体的语法规则参见 [Hint 语法](#语法)部分。
+
+### 排序规则不同导致 `INL_JOIN` Hint 不生效
+
+如果两个表的 Join key 的排序规则设置不同，将无法使用 IndexJoin 来执行查询。此时 [`INL_JOIN` Hint](#inl_joint1_name--tl_name-) 将无法生效。例如：
```
[the remark this answers is not in the document]
Comment (qw4990) [reply optimizer-hints.md]: 已更新

**Review by time-and-fate — APPROVED** [review]

File: optimizer-hints.md [hunk 21]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
 SELECT /*+ RESOURCE_GROUP(rg1) */ * FROM t limit 10;
 ```
+
+## 常见 Hint 不生效问题排查
+
+### MySQL 命令行客户端清除 Hint 导致不生效
+
+MySQL 命令行客户端在 5.7.7 版本之前默认清除了 Optimizer Hints。如果需要在这些早期版本的客户端中使用 Hint 语法，需要在启动客户端时加上 `--comments` 选项。例如 `mysql -h 127.0.0.1 -P 4000 -uroot --comments`。
+
+### 不指定库名导致 Hint 不生效
```
Comment (qiancai) [inline optimizer-hints.md:796]:
```suggestion
### 创建连接时不指定库名导致 Hint 不生效
```

Comment (qiancai) [reply optimizer-hints.md:796]: reason: 避免觉得原来的标题有重合关系
- 不指定库名导致 Hint 不生效
- 跨库查询不指定库名导致 Hint 不生效

File: optimizer-hints.md [hunk 22]
```diff
@@ -814,3 +786,156 @@ SELECT /*+ NTH_PLAN(3) */ count(*) from t where a > 5;
...
+由于 `t1` 不在当前数据库 `test2` 下，因此 `use_index(t1, a)` Hint 无法被正确地识别。此时，需要显式地指定库名，即修改为 `use_index(test1.t1, a)`。
+
+```sql
++---------+------+----------------------------------------------------------------------------------+
+| Level   | Code | Message                                                                          |
++---------+------+----------------------------------------------------------------------------------+
+| Warning | 1815 | use_index(test2.t1, a) is inapplicable, check whether the table(test2.t1) exists |
++---------+------+----------------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
```
Comment (qiancai) [inline optimizer-hints.md:840]:
````suggestion
由于 `t1` 不在当前数据库 `test2` 下，因此 `use_index(t1, a)` Hint 无法被正确地识别。

```sql
+---------+------+----------------------------------------------------------------------------------+
| Level   | Code | Message                                                                          |
+---------+------+----------------------------------------------------------------------------------+
| Warning | 1815 | use_index(test2.t1, a) is inapplicable, check whether the table(test2.t1) exists |
+---------+------+----------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

此时，需要显式地指定库名，即将 `use_index(t1, a)` 修改为 `use_index(test1.t1, a)`。
````

**Review by chrysan — APPROVED** [review]

**Review by qiancai — APPROVED** [review]
/lgtm

**Pull request merged**